src/staking/index.ts

```diff
@@ -0,0 +1,121 @@
+import { StakingPoolFactory__factory, StakingPool__factory } from "@energyweb/iam-contracts";
+import { StakingPool } from "@energyweb/iam-contracts/dist/ethers-v4/StakingPool";
+import { StakingPoolFactory } from "@energyweb/iam-contracts/dist/ethers-v4/StakingPoolFactory";
+import { utils } from "ethers";
+import { BigNumber } from "ethers/utils";
+import { ERROR_MESSAGES } from "../errors";
+import { IAM } from "../iam";
+import { chainConfigs } from "../iam/chainConfig";
+import { EncodedCall } from "../iam/iam-base";
+
+const { namehash } = utils;
+/**
+ * @interface
+ */
+export interface IStakable {
+  /**
+     * @description Deployes organization staking pool
+     * @emits StakingPoolFactory.StakingPoolLaunched
+     */
+  launchStakingPool(
+    { org, minStakingPeriod, patronRewardPortion, patronRoles, principal }:
+      { org: string, minStakingPeriod: number | BigNumber, patronRewardPortion: number, patronRoles: string[], principal: BigNumber }
+  ): Promise<void>;
+  /**
+   * @description Locks stake and starts accumulating reward
+   * @emits StakingPool.StakePut
+   */
+  putStake(
+    { stake, org }:
+      { stake: BigNumber, org: string }
+  ): Promise<void>;
+  /**
+   * @param org ENS name of organization
+   */
+  getPool(org: string): Promise<StakingPool>;
+  /**
+    * @description Stops accumulating of the reward and prepars stake to withdraw after withdraw delay. 
+    * Withdraw request unavailable until minimum staking period ends
+    * @param org - ENS name of organization
+    */
+  requestWithdraw(org: string): Promise<void>;
+  /**
+   * @description pays back stake with accumulated reward. Withdrawn unavailable until withdrawn delay ends
+   * @param org ENS name of organization
+   * @emits StakingPool.StakeWithdrawn
+   */
+  withdraw(org: string): Promise<void>;
+}
+
+type Constructor<T = unknown> = new (...args: any[]) => T;
+/**
+ * @description implements IStakable on top of `base`
+ * 
+ * @param base - constructor of any IAM-superclass
+ * 
+ * @returns constructor of class extending `IAM` and implementing IStakable
+ */
+export function Stakable<T extends Constructor<IAM>>(base: T): new (...args: ConstructorParameters<typeof IAM>) => IAM & IStakable {
+  return class extends base implements IStakable {
+    private stakingPoolFactory: StakingPoolFactory;
```

Comment:
It's not immediately obvious to me why this approach of a function that returns a constructor is being used instead of basic JS class inheritance. Is there some pattern that is being implemented here?

Reply:
My goal is eventually to separate ENS, Roles and Asset functionality. Class with all these mixins will have constructor `ENS(Roles(Assets(Stakable(IAM))))`. Using inheritance for this would require creation of classes `Stakable extends IAM`, `Assets extends Stakable`, `Roles extends Assets`, `ENS extends Roles` of which only latter would be used. Although inheritance approach is completely doable, from object oriented design point of view there are no `is` relationship between these classes